### PR TITLE
Add hello-xgboost example

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
- Remove hello-mpi example
- Add hello-xgboost example
- Add app_opt.xgboost module based on @nvidianz 's original design
- Adopt Rong's code

### CODE REVIEW

File: nvflare/app_opt/xgboost/controller.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        bin_folder = workspace.get_startup_kit_dir()
+        server_cert_path = os.path.join(bin_folder, "server.crt")
+        if not os.path.exists(server_cert_path):
+            self.log_error(fl_ctx, "Missing server certificate (server.crt)")
+            return False
+        server_key_path = os.path.join(bin_folder, "server.key")
+        if not os.path.exists(server_key_path):
+            self.log_error(fl_ctx, "Missing server key (server.key)")
+            return False
+        client_cert_path = os.path.join(bin_folder, "rootCA.pem")
```
Comment (nvidianz) [inline nvflare/app_opt/xgboost/controller.py:70]: This should be called ca_cert_path

File: nvflare/app_opt/xgboost/controller.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        rank_map = {clients[i].name: i for i in range(0, len(clients))}
+        self._rank_map = rank_map
+        self._participate_clients = clients
+
+        if not self._port:
+            self._port = get_open_ports(1)[0]
+
+        self.log_info(fl_ctx, f"Starting XGBoost FL server on port {self._port}")
+
+        if not self._get_certificates(fl_ctx):
```
Comment (nvidianz) [inline nvflare/app_opt/xgboost/controller.py:93]: This should be in sync with NVFlare mode. If NVFlare is in secure mode, certificate should be required.

File: nvflare/app_opt/xgboost/controller.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    def stop_controller(self, fl_ctx: FLContext):
+        self.cancel_all_tasks()
+        if self._xgb_fl_server:
+            self._xgb_fl_server.terminate()
+
+    def process_result_of_unknown_task(
+        self, client: Client, task_name, client_task_id, result: Shareable, fl_ctx: FLContext
+    ):
+        self.log_warning(fl_ctx, f"Unknown task: {task_name} from client {client.name}.")
```
Comment (nvidianz) [inline nvflare/app_opt/xgboost/controller.py:114]: This should be an error.

File: nvflare/app_opt/xgboost/learner_base.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        """Main XGBoost training method"""
+        pass
+
+    def initialize(self, parts: dict, fl_ctx: FLContext):
+        self.client_id = fl_ctx.get_identity_name()
+
+    def _get_certificates(self, fl_ctx: FLContext):
+        workspace: Workspace = fl_ctx.get_prop(FLContextKey.WORKSPACE_OBJECT)
+        bin_folder = workspace.get_startup_kit_dir()
+        server_cert_path = os.path.join(bin_folder, "rootCA.pem")
```
Comment (nvidianz) [inline nvflare/app_opt/xgboost/learner_base.py:92]: This should be called ca_cert_path

**Review by nvidianz — APPROVED** [review]

File: nvflare/app_opt/xgboost/controller.py (hunk 1 above)
Comment (YuanTingHsieh) [reply nvflare/app_opt/xgboost/controller.py]: thanks changed.

File: nvflare/app_opt/xgboost/controller.py (hunk 3 above)
Comment (YuanTingHsieh) [reply nvflare/app_opt/xgboost/controller.py]: thanks changed.

File: nvflare/app_opt/xgboost/learner_base.py (hunk 4 above)
Comment (YuanTingHsieh) [reply nvflare/app_opt/xgboost/learner_base.py]: thanks changed.

File: nvflare/app_opt/xgboost/controller.py (hunk 2 above)
Comment (YuanTingHsieh) [reply nvflare/app_opt/xgboost/controller.py]: thanks, I've fixed it.

**Review by nvidianz — APPROVED** [review]

File: nvflare/app_opt/xgboost/learner.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    def xgb_train(self, params: XGBoostParams, fl_ctx: FLContext) -> Shareable:
+        xgb, flag = optional_import(module="xgboost")
+        if not flag:
+            self.log_error(fl_ctx, "Can't import xgboost")
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+        callback, flag = optional_import(module="xgboost.callback")
+        if not flag:
+            self.log_error(fl_ctx, "Can't import xgboost.callback")
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+        with xgb.rabit.RabitContext([e.encode() for e in params.rabit_env]):
```
Comment (rongou) [inline nvflare/app_opt/xgboost/learner.py:34]: Seems like everything before this line could be in the base?

File: nvflare/app_opt/xgboost/learner.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                params.xgb_params,
+                dtrain,
+                params.num_rounds,
+                evals=watchlist,
+                early_stopping_rounds=params.early_stopping_rounds,
+                verbose_eval=params.verbose_eval,
+                callbacks=[callback.EvaluationMonitor(rank=self.rank)],
+            )
+
+            # Save the model.
```
Comment (rongou) [inline nvflare/app_opt/xgboost/learner.py:54]: Is there a standard/recommended way to saving models in nvflare? Perhaps this can also be in the base?

**Pull request merged**